File: typescript/src/sorting/mergeSort.ts
export function merge(arr: number[], l: number, m: number, r: number): void{
    let sizeLeft = m - l + 1;
    let sizeRight = r - m;

    let left: number[] = new Array(sizeLeft);
    let right: number[] = new Array(sizeRight);

    for (let i = 0; i < sizeLeft; i++) {
        left[i] = arr[l + i]
        
    }
    for (let j = 0; j < sizeRight; j++) {
        right[j] = arr[m + j + 1];
    }

    let i = 0, j = 0;
    let k = l;

    while (i < sizeLeft && j < sizeRight) {
        if (left[i] < right[j]) {
            arr[k] = left[i];
            i++;
        } else {
            arr[k] = right[j];
            j++;
        }
        k++;
    }

    while (i < sizeLeft) {
        arr[k] = left[i];
        i++;
        k++
    }
    while (j < sizeRight) {
        arr[k] = right[j];
        j++;
        k++;
    }
}

export function sort(arr: number[], l: number, r: number): void{
    if (l < r) {
        let middle = l + Math.floor((r-l)/2);

        sort(arr, l, middle);
        sort(arr, middle + 1, r);

        merge(arr, l, middle, r);
    }
}

export function mergeSort(arr: number[]): void{
    sort(arr,0,arr.length-1);
}

File: python/sorting/merge_sort.py

def merge_sort(arr: list[int]):
    if len(arr) > 1:
        mid = len(arr)//2

        left = arr[:mid]
        right = arr[mid:]

        merge_sort(left)
        merge_sort(right)

        i = j = k = 0

        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                arr[k] = left[i]
                i += 1
            else:
                arr[k] = right[j]
                j += 1

            k += 1
        

        while i < len(left):
            arr[k] = left[i];
            i += 1
            k += 1

        while j < len(right):
            arr[k] = right[j]
            j += 1
            k += 1


def test_merge_sort():
    input = [12, 222, 15, 76, 43, 500, 23, 8, 65, 32, 69, 420, 344, 6754]
    expected = [8, 12, 15, 23, 32, 43, 65, 69, 76, 222, 344, 420, 500, 6754]

    merge_sort(input)

    assert input == expected, "List unsorted"

